tvm/src/codegen/llvm/codegen_llvm.cc

```diff
@@ -1515,7 +1525,42 @@ llvm::Value* CodeGenLLVM::VisitExpr_(const KernelExpr* op) {
     arg_types.push_back(LLVMType(arg.type()));
   }
   llvm::Function* f = module_->getFunction(op->name);
-  return builder_->CreateCall(f, arg_value);
+  llvm::Value* ret_val = builder_->CreateCall(f, arg_value);
+  llvm::BasicBlock* assert_true_kernel =
+      llvm::BasicBlock::Create(*ctx_, "assert_true_kernel", function_);
+  llvm::BasicBlock* assert_false_kernel =
+      llvm::BasicBlock::Create(*ctx_, "assert_false_kernel", function_);
+  llvm::Value* cond =
+      builder_->CreateLoad(llvm::Type::getInt32Ty(*ctx_), assert_global_ptr_);
+  builder_->CreateCondBr(cond, assert_true_kernel, assert_false_kernel);
+  builder_->SetInsertPoint(assert_false_kernel);
+
+  std::string if_then_name = "if_then_";
+  std::string if_end_name = "if_end_";
+  from_assert_ = true;
+  for (size_t num_free = 0; num_free < assert_alloc_mem_.size(); num_free++) {
+    Expr free_op =
+        Call::make(Int(32), "TVMBackendFreeWorkspace",
+                   {cast(Int(32), assert_alloc_mem_[num_free].dev_type_),
+                    cast(Int(32), assert_alloc_mem_[num_free].dev_id_),
+                    assert_alloc_mem_[num_free].buffer_var_},
+                   Call::Extern);
+    llvm::Value* v = MakeValue(free_op);
+    llvm::Value* ne = builder_->CreateICmpNE(v, ConstInt32(0));
+    llvm::BasicBlock* if_true_ = llvm::BasicBlock::Create(
+        *ctx_, if_then_name + std::to_string(num_free), function_);
+    llvm::BasicBlock* if_end_ = llvm::BasicBlock::Create(
+        *ctx_, if_end_name + std::to_string(num_free), function_);
+    builder_->CreateCondBr(ne, if_end_, if_true_);
+    builder_->SetInsertPoint(if_end_);
+    builder_->CreateRet(ConstInt32(-1));
+    builder_->CreateBr(if_true_);
+    builder_->SetInsertPoint(if_true_);
+  }
```

Comment:
We should put this logic in a separate module and reuse that in your assert statement.